realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -433,22 +435,83 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeGetTable(JNIEnv
 
 JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeCreateTable(JNIEnv* env, jclass,
                                                                              jlong shared_realm_ptr,
-                                                                             jstring table_name)
+                                                                             jstring table_name,
+                                                                             jboolean is_pk_table)
 {
     TR_ENTER_PTR(shared_realm_ptr)
 
-    std::string name_str;
+    std::string table_name_str;
     try {
-        JStringAccessor name(env, table_name); // throws
-        name_str = name;
+        table_name_str = JStringAccessor(env, table_name); // throws
         auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
         shared_realm->verify_in_write(); // throws
-        Table* table = LangBindHelper::add_table(shared_realm->read_group(), name); // throws
+        Table* table;
+        auto& group = shared_realm->read_group();
+        if (is_pk_table) {
+            // sync::create_table() will add an extra column for stable ID which is not allowed for pk table.
+            table = LangBindHelper::add_table(group, table_name_str); // throws
+        }
+        else {
+#if REALM_ENABLE_SYNC
+            // Sync doesn't throw when table exists.
+            if (group.has_table(table_name_str)) {
+                THROW_JAVA_EXCEPTION(
+                    env, JavaExceptionDef::IllegalArgument,
+                    format(c_table_name_exists_exception_msg, table_name_str.substr(TABLE_PREFIX.length())));
+            }
+            auto table_ref = sync::create_table(group, table_name_str); // throws
+            table = LangBindHelper::get_table(group, table_ref->get_index_in_group());
+#else
+            table = LangBindHelper::add_table(group, table_name_str); // throws
+#endif
+        }
+        return reinterpret_cast<jlong>(table);
+    }
+    catch (TableNameInUse& e) {
+        // We need to print the table name, so catch the exception here.
+        std::string class_name_str(table_name_str.substr(TABLE_PREFIX.length()));
+        ThrowException(env, IllegalArgument, format(c_table_name_exists_exception_msg, class_name_str));
+    }
+    CATCH_STD()
+
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeCreateTableWithPrimaryKeyField(
+    JNIEnv* env, jclass, jlong shared_realm_ptr, jstring table_name, jstring field_name, jboolean is_string_type,
+    jboolean is_nullable)
+{
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    std::string class_name_str;
+    try {
+        std::string table_name_str(JStringAccessor(env, table_name));
```

Comment:
We really need to find a common pattern for the naming of java parameters that get converted to c++ ones and then document it. Here we use `field_name` -> `field_name_str`. In other places it is `fieldName` -> `field_name` and yet others it is `j_field_name` -> `field_name`. I don't particularly care which to use, although I would probably prefer if the c++ name was as short as possible i.e. `field_name`.